### DISCUSSION TITLE
ReactElement in label is not restoring correctly using setNodes

### DISCUSSION THREAD

**Discussion by ShaneHudson** [xyflow/xyflow#2728] [General]
I'm unsure whether this is a bug in react-flow or something I need to do myself. Essentially ReactElement works nicely in Nodes, so we can put whatever JSX we want in and it will work... this is great and exactly what I need.

However when using the save/restore example from the docs (using `.toObject()` and saving the JSON as a string), it fails on the restore due to the classic old "use array not object" error for the ReactElements.

So my question is, does react-flow need a bugfix to enable `setNodes` to parse the ReactElement when parsed from JSON?

**Comment by moklick** [comment]
This was a bad API decision from our side. You can't serialize React components that easy, that's why this is failing. You should better implement a custom node and only store serializable values in your node data objects.

**Comment by ShaneHudson** [comment]
Ah I understand, so provide a model that gets stored and then passed into a component that can render it properly. Got it.

Thank you so much!

**Reply by moklick** [reply]
Yep! That's way better :)
